relayer/chain/parachain/listener.go

```diff
@@ -0,0 +1,272 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package parachain
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/sirupsen/logrus"
+	merkletree "github.com/wealdtech/go-merkletree"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/snowfork/go-substrate-rpc-client/v2/scale"
+	"github.com/snowfork/go-substrate-rpc-client/v2/types"
+	"github.com/snowfork/polkadot-ethereum/relayer/chain"
+	chainTypes "github.com/snowfork/polkadot-ethereum/relayer/substrate"
+)
+
+type Listener struct {
+	config   *Config
+	conn     *Connection
+	messages chan<- []chain.Message
+	log      *logrus.Entry
+}
+
+func NewListener(config *Config, conn *Connection, messages chan<- []chain.Message, log *logrus.Entry) *Listener {
+	return &Listener{
+		config:   config,
+		conn:     conn,
+		messages: messages,
+		log:      log,
+	}
+}
+
+func (li *Listener) Start(ctx context.Context, eg *errgroup.Group) error {
+	li.log.Infof("\n\nParachain listener empty start function\n\n")
+
+	eg.Go(func() error {
+		return li.subBeefyJustifications(ctx)
+	})
+
+	return nil
+}
+
+func (li *Listener) onDone(ctx context.Context) error {
+	li.log.Info("Shutting down listener...")
+	close(li.messages)
+	return ctx.Err()
+}
+
+func (li *Listener) subBeefyJustifications(ctx context.Context) error {
+	ch := make(chan interface{})
+
+	sub, err := li.conn.api.Client.Subscribe(context.Background(), "beefy", "subscribeJustifications", "unsubscribeJustifications", "justifications", ch)
+	if err != nil {
+		panic(err)
+	}
+	defer sub.Unsubscribe()
+
+	// timeout := time.After(40 * time.Second)
+	received := 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			return li.onDone(ctx)
+		case msg := <-ch:
+			li.log.Info("encoded msg: %#v\n", msg)
+
+			// TODO: var messages []chainTypes.BeefyMessage
+			signedCommitment := &SignedCommitment{}
+			err := types.DecodeFromHexString(msg.(string), signedCommitment)
+			if err != nil {
+				li.log.WithError(err).Error("Faild to decode beefy commitment messages")
+			}
+
+			received++
+			li.log.Info("Msg number:", received)
+
+			// Construct BEEFY merkle tree
+			beefyValidatorAddresses := []string{
```

Comment:
are you having in issues querying the authorities from the chain via beefy.authorities? lmk if stuck

Reply:
Included this as part of issue https://github.com/Snowfork/polkadot-ethereum/issues/341